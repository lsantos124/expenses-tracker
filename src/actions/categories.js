import database from '../firebase/firebase';

// ADD_CATEGORY

export const addCategory = (category) => ({
	type: 'ADD_CATEGORY',
	category
});

export const startAddCategory = (categoryData = {}) => {
	return (dispatch, getState) => {
        const uid = getState().auth.uid;
		const {
            title = '',
			description = '',
			color = '' 
        } = categoryData;

		const category = { title, description, color };

		return database.ref(`users/${uid}/categories`).push(category).then((ref) => {
			dispatch(addCategory({
				id: ref.key,
				...category
			}));
		});
	};
};

// REMOVE CATEGORY

export const removeCategory = ({ id } = {}) => ({
	type: 'REMOVE_CATEGORY',
	id
});

export const startRemoveCategory = ({ id } = {}) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		return database.ref(`users/${uid}/categories/${id}`).remove().then(() => {
			dispatch(removeCategory({ id }));
		});
	};
};

// SET CATEGORY

export const setCategories = (categories) => ({
	type: 'SET_CATEGORIES',
	categories
});

export const startSetCategories = () => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		return database.ref(`users/${uid}/categories`)
			.once('value')
			.then((snapshot) => {
				const categories = [];

				snapshot.forEach((childSnapshot) => {
					categories.push({
						id: childSnapshot.key,
						...childSnapshot.val()
					});
				});

				dispatch(setCategories(categories));
			});
	};
};