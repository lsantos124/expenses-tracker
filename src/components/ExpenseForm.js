import React from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import { SingleDatePicker } from 'react-dates';

export class ExpenseForm extends React.Component {
	constructor(props) {
		super(props);
	
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			category: props.expense ? props.expense.category : '',
			calendarFocused: false,
			error: ''
		};
	}

	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	};

	onNoteChange = (e) => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	}

	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	}

	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));			
		}
	};

	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));
	};

	onSelectChange = (e) => {
		const category = e.target.value;
		this.setState(() => ({ category }));			
	};

	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({ error: 'Please provide description and amount.' }));
		} else {
			this.setState(() => ({ error: '' }));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note,
				category: this.state.category
			});
		}
	};

	render() {
		return (
			<form className="form" onSubmit={this.onSubmit}>
				{ this.state.error && <p className="form__error">{this.state.error}</p>}
				<input
					type="text"
					placeholder="Description"
					autoFocus
					className="text-input"
					value={this.state.description}
					onChange={this.onDescriptionChange}
				/>
				<input
					type="text"
					className="text-input"
					placeholder="Amount"
					value={this.state.amount}
					onChange={this.onAmountChange}
				/>
				<SingleDatePicker
					date={this.state.createdAt}
					onDateChange={this.onDateChange}
					focused={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
				<div>
					<label className="select__label">Category:</label>
					<select 
						className="select"
						defaultValue={this.state.category}
						onChange={this.onSelectChange}
					>
						<option value="">None</option>
						{
							this.props.categories.map(({id, title, color}) => {
								return (
									<option
										key={id} 
										value={title}
									>
										{title}
									</option>
								);
							})
						}
					</select>
				</div>
				<textarea
					className="textarea"
					placeholder="Add a note for your expense (optional)"
					value={this.state.note}
					onChange={this.onNoteChange}
				>
				</textarea>
				<div>
					<button className="button">Save Expense</button>
				</div>
			</form>
		);
	}
}

const mapStateToProps = (state) => ({
	categories: state.categories
});

export default connect(mapStateToProps)(ExpenseForm);