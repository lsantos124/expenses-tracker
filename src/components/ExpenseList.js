import React from 'react';
import { connect } from 'react-redux';
import { CSVLink } from 'react-csv';
import moment from 'moment';
import numeral from 'numeral';
import ExpenseListItem from './ExpenseListItem';
import { getVisibleExpenses } from '../selectors/expenses';

export const ExpenseList = (props) => {
	const csvHeaders = [['description', 'amount', 'createdAt']];
	const csvExpenseData = props.expenses.map(({ id, description, amount, createdAt }) => {
		const createdAtFormatted = moment(createdAt).format('MMMM Do, YYYY');
		const amountFormatted = numeral(amount / 100).format('$0,0.00');
		return [description, amountFormatted, createdAtFormatted];
	});
	const csvData = [ ...csvHeaders, ...csvExpenseData ];

	const startDateFormatted = moment(props.startDate).format('MMMM Do, YYYY');
	const endDateFormatted = moment(props.endDate).format('MMMM Do, YYYY');
	const fileName = `text=${props.filters.text}_startDate=${startDateFormatted}_endDate=${endDateFormatted}.csv`;
	
	return (
		<div className="content-container">
			<div className="list-header">
				<div className="show-for-mobile">Expenses</div>
				<div className="show-for-desktop">Expense</div>
				<div className="show-for-desktop">Amount</div>
			</div>
			<div className="list-body">
			{
				props.expenses.length === 0 ? (
					<div className="list-item list-item--message">
						<span>No expenses</span>
					</div>
				) : (
					props.expenses.map((expense) => {
						let categoryColor = '';
						props.categories.map((category) => {
							if (category.title == expense.category) {
								categoryColor = category.color;
							}
							return category
						})

						return <ExpenseListItem key={expense.id} {...expense} color={categoryColor} />;
					})
				)
			}
			</div>
			<div className="list-buttons">
				<CSVLink 
					className="button button--csvlink" 
					data={csvData}
					filename={fileName}
					>
						Download
					</CSVLink>
			</div>
		</div>
	);
};

const mapStateToProps = (state) => {
	return {
		expenses: getVisibleExpenses(state.expenses, state.filters),
		filters: state.filters,
		categories: state.categories
	};
}

export default connect(mapStateToProps)(ExpenseList);