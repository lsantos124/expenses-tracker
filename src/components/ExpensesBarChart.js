// MyBarChart.js
import React from 'react';
import {
    FlexibleWidthXYPlot,
    XAxis, // Shows the values on x axis
    YAxis, // Shows the values on y axis
    VerticalBarSeries,
    LabelSeries,
    HorizontalGridLines
} from 'react-vis';

class ExpensesBarChart extends React.Component {
    render() {
    	let maxData = 0;

		this.props.data.map((data) => {
			maxData = Math.max(maxData, data['y']);
			return data;
		});
        
        const data = this.props.data;
        const chartHeight = 250;
        const chartDomain = [0, maxData + 20];
        return (
            <FlexibleWidthXYPlot 
                xType="ordinal" 
                height={chartHeight}
                yDomain={chartDomain}
            >
                <XAxis 
                	title="Date"
                	tickLabelAngle={45} 
                	tickPadding={35}
            	    tickValues={(
            	    	data.length > 15)
			              ? data
			                .filter((item, idx) => {
			                  if ((idx % Math.floor(data.length / 15)) === 0) {
			                    return item.x
			                  }
			                }).map(item => (item.x))
			              : data.map(item => (item.x))
			          }
			    />
                <YAxis title="Amount ($)"/>
                <VerticalBarSeries
                    data={data}
                />
                <LabelSeries
                    data={data.map(obj => {
                        return { ...obj, label: obj.y.toString() }
                    })}
                    labelAnchorX="start"
                    labelAnchorY="text-after-edge"
                    rotation={-45}
                />
                <HorizontalGridLines />
            </FlexibleWidthXYPlot>
        );
    }
}
export default ExpensesBarChart;