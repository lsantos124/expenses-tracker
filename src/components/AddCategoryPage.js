import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { TwitterPicker } from 'react-color';
import { startAddCategory } from '../actions/categories';

export class AddCategoryPage extends React.Component {
    constructor(props) {
		super(props);
	
		this.state = {
			title: '',
			description: '',
			color: '',
			error: ''
		};
    }
    
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    handleChangeComplete = (color, event) => {
        this.setState({ color: color.hex });
    };
    
    onSubmit = (e) => {
        e.preventDefault();
        
		if (!this.state.title) {
			this.setState({ error: 'Please provide category title' });
		}
		else {
			let categoryExists = false;
			this.props.categories.map((category) => {
				if (category.title == this.state.title) {
					categoryExists = true;
				}
				return category;
			})

			if (categoryExists) {
				this.setState({ error: 'Category already exists!' });
			} else {
				this.props.startAddCategory({
					title: this.state.title,
					description: this.state.description,
					color: this.state.color
				});
                this.props.history.push('/customize');
			}
		}
    }
    
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Category</h1>
                    </div>
                </div>
                <div className="content-container">
                    <form className="form" onSubmit={this.onSubmit}>
                        { this.state.error && <p className="form__error">{this.state.error}</p> }
                        <input 
                            type="text"
                            className="text-input" 
                            placeholder="Category Title"
                            value={this.state.title} 
                            onChange={this.onTitleChange}
                        />
                        <input 
                            type="textarea" 
                            className="textarea"
                            placeholder="Category Description"
                            value={this.state.description} 
                            onChange={this.onDescriptionChange} 
                        />
                        <TwitterPicker 
                            color={this.state.color} 
                            onChangeComplete={this.handleChangeComplete} 
                        />
                        <button className="button">Add Category</button>
                    </form>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
	categories: state.categories
});

const mapDispatchToProps = (dispatch) => ({
    startAddCategory: (category) => dispatch(startAddCategory(category))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddCategoryPage);