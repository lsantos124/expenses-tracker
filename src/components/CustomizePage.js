import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startRemoveCategory } from '../actions/categories';

export const CustomizePage = ({ categories, startRemoveCategory }) => {
	return (
		<div>
			<div className="page-header">
				<div className="content-container">
					<h1 className="page-header__title">Customize Settings</h1>
				</div>
			</div>
			<div className="content-container">
				<div className="list-header">
					<div>Categories</div>
				</div>
				<div className="list-body">
				{
					categories.map((category, index) => {
						return (
							<div className="list-item" key={index}>
								<div className="list-item__color">
									<div className="box-color" style={{backgroundColor: category.color}}></div>
									<div>
										<h3 className="list-item__title">{category.title}</h3>
										<span className="list-item__subtitle">{category.description}</span>
									</div>
								</div>
								<button 
									className="button button--secondary"
									onClick={() => startRemoveCategory({ id: category.id })}
									>
										Delete
									</button>
							</div>
						);
					})
				}
				</div>
				<div className="list-buttons">
					<Link 
						to='/addCategory'
						className="button"
					>
						Add Category
					</Link>
				</div>
				
			</div>
		</div>
	)
};

const mapStateToProps = (state) => ({
	categories: state.categories
});

const mapDispatchToProps = (dispatch) => ({
	startRemoveCategory: ({id}) => dispatch(startRemoveCategory({id}))
});

export default connect(mapStateToProps, mapDispatchToProps)(CustomizePage);