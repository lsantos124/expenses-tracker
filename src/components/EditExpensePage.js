import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import RemoveModal from './RemoveModal';
import { startRemoveExpense, startEditExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
	state = {
		delete: false
	}

	onSubmit = (expense) => {
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	}

	onClick = () => {
		this.setState(() => ({ delete: true }));	
	}

	handleDeleteExpense = () => {
		this.setState(() => ({ delete: false }));
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/');	
	};

	handleCancelDeleteExpense = () => {
		this.setState(() => ({ delete: false }));	
	};

	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Edit Expense</h1>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm 
						expense={this.props.expense}
						onSubmit={this.onSubmit}
					/>
					<button className="button button--secondary" onClick={this.onClick}>
						Remove Expense
					</button>
				</div>
				<RemoveModal 
					delete={this.state.delete}
					handleDeleteExpense={this.handleDeleteExpense}
					handleCancelDeleteExpense={this.handleCancelDeleteExpense}
				/>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
	};
};

const mapDispatchToProps = (dispatch) => {
	return {
		startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
		startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);