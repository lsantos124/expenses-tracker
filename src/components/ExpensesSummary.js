import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { getVisibleExpenses, getHiddenExpenses } from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import getExpensesOverTime from '../selectors/expenses-over-time';
import numeral from 'numeral';
import ExpensesBarChart from './ExpensesBarChart';

export const ExpensesSummary = ({ expenseCount, expensesTotal, hiddenExpenseCount, expensesOverTime }) => {
	const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
	const hiddenExpenseWord = hiddenExpenseCount === 1 ? 'expense' : 'expenses';
	const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

	return (
		<div className="page-header">
			<div className="content-container">
				<h1 className="page-header__title">Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span></h1>
				<h3 className="page-header__subtitle"><span>{hiddenExpenseCount}</span> {hiddenExpenseWord} hidden</h3>
				<div className="page-header__actions">
					<Link className="button" to="/create">Add Expense</Link>
				</div>
				<div className="graph">
					<ExpensesBarChart data={expensesOverTime}/>
				</div>
			</div>
		</div>
	);
};

const mapStateToProps = (state) => {
	const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
	const hiddenExpenses = getHiddenExpenses(state.expenses, state.filters);

	return {
		expenseCount: visibleExpenses.length,
		expensesTotal: getExpensesTotal(visibleExpenses),
		hiddenExpenseCount: hiddenExpenses.length,
		expensesOverTime: getExpensesOverTime(visibleExpenses, state.filters)
	};
};

export default connect(mapStateToProps)(ExpensesSummary);