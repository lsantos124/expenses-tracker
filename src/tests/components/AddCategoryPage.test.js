import React from 'react';
import { shallow } from 'enzyme';
import { AddCategoryPage } from '../../components/AddCategoryPage';
import categories from '../fixtures/categories';

test('should render AddCategoryPage correctly', () => {
    const wrapper = shallow(<AddCategoryPage categories={categories}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should set title on input change', () => {
    const value = categories[0].title;
    const wrapper = shallow(<AddCategoryPage categories={categories}/>);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('title')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('should set description on input change', () => {
    const value = categories[0].description;
    const wrapper = shallow(<AddCategoryPage categories={categories}/>);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('description')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    const value = 'new title';
    const history = { push: jest.fn() };
    const startAddCategory = jest.fn();
    const wrapper = shallow(<AddCategoryPage categories={categories} history={history} startAddCategory={startAddCategory}/>);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('title')).toBe(value);

    const e = { preventDefault: jest.fn() };
    wrapper.find('form').prop('onSubmit')(e);
	expect(history.push).toHaveBeenLastCalledWith('/customize');
	expect(startAddCategory).toHaveBeenLastCalledWith({
        title: value,
        description: '',
        color: ''
    });
});

test('should generate error message if no category provided', () => {
    const wrapper = shallow(<AddCategoryPage categories={categories}/>);
    const e = { preventDefault: jest.fn() };
    wrapper.find('form').prop('onSubmit')(e);
	expect(wrapper.state('error')).toBe('Please provide category title');
});

test('should generate error message if category exists', () => {
    const value = categories[0].title;
    const history = { push: jest.fn() };
    const startAddCategory = jest.fn();
    const wrapper = shallow(<AddCategoryPage categories={categories} history={history} startAddCategory={startAddCategory}/>);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('title')).toBe(value);

    const e = { preventDefault: jest.fn() };
    wrapper.find('form').prop('onSubmit')(e);
    expect(wrapper.state('error')).toBe('Category already exists!');
});