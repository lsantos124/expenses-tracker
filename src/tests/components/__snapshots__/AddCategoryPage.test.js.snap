// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddCategoryPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Category
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Category Title"
        type="text"
        value=""
      />
      <input
        className="textarea"
        onChange={[Function]}
        placeholder="Category Description"
        type="textarea"
        value=""
      />
      <ColorPicker
        color=""
        colors={
          Array [
            "#FF6900",
            "#FCB900",
            "#7BDCB5",
            "#00D084",
            "#8ED1FC",
            "#0693E3",
            "#ABB8C3",
            "#EB144C",
            "#F78DA7",
            "#9900EF",
          ]
        }
        onChangeComplete={[Function]}
        styles={Object {}}
        triangle="top-left"
        width={276}
      />
      <button
        className="button"
      >
        Add Category
      </button>
    </form>
  </div>
</div>
`;

exports[`should set description on input change 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Category
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Category Title"
        type="text"
        value=""
      />
      <input
        className="textarea"
        onChange={[Function]}
        placeholder="Category Description"
        type="textarea"
        value=""
      />
      <ColorPicker
        color=""
        colors={
          Array [
            "#FF6900",
            "#FCB900",
            "#7BDCB5",
            "#00D084",
            "#8ED1FC",
            "#0693E3",
            "#ABB8C3",
            "#EB144C",
            "#F78DA7",
            "#9900EF",
          ]
        }
        onChangeComplete={[Function]}
        styles={Object {}}
        triangle="top-left"
        width={276}
      />
      <button
        className="button"
      >
        Add Category
      </button>
    </form>
  </div>
</div>
`;

exports[`should set title on input change 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Category
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Category Title"
        type="text"
        value="Charity"
      />
      <input
        className="textarea"
        onChange={[Function]}
        placeholder="Category Description"
        type="textarea"
        value=""
      />
      <ColorPicker
        color=""
        colors={
          Array [
            "#FF6900",
            "#FCB900",
            "#7BDCB5",
            "#00D084",
            "#8ED1FC",
            "#0693E3",
            "#ABB8C3",
            "#EB144C",
            "#F78DA7",
            "#9900EF",
          ]
        }
        onChangeComplete={[Function]}
        styles={Object {}}
        triangle="top-left"
        width={276}
      />
      <button
        className="button"
      >
        Add Category
      </button>
    </form>
  </div>
</div>
`;
